File: Services/UserCacheManager.swift
import Foundation
import UIKit

class UserCacheManager {
    static let shared = UserCacheManager()
    
    private let userDefaults = UserDefaults.standard
    private let cacheKey = "cached_user_data"
    private let cacheTimestampKey = "cached_user_timestamp"
    private let profileImageCacheKey = "cached_profile_image"
    private let partnerImageCacheKey = "cached_partner_image"
    private let partnerImageURLKey = "cached_partner_image_url"
    
    private init() {}
    
    // MARK: - Cache User Data
    
    /// Cache l'utilisateur après chargement Firebase
    func cacheUser(_ user: AppUser) {
        print("💾 UserCacheManager: Mise en cache utilisateur: \(user.name)")
        
        do {
            let encoder = JSONEncoder()
            let data = try encoder.encode(user)
            userDefaults.set(data, forKey: cacheKey)
            userDefaults.set(Date().timeIntervalSince1970, forKey: cacheTimestampKey)
            
            print("✅ UserCacheManager: Utilisateur mis en cache avec succès")
        } catch {
            print("❌ UserCacheManager: Erreur encodage utilisateur: \(error)")
        }
    }
    
    /// Récupère l'utilisateur en cache (instantané)
    func getCachedUser() -> AppUser? {
        guard let data = userDefaults.data(forKey: cacheKey) else {
            print("📂 UserCacheManager: Aucun utilisateur en cache")
            return nil
        }
        
        // Vérifier l'âge du cache (optionnel - pour éviter des données trop anciennes)
        if let timestamp = userDefaults.object(forKey: cacheTimestampKey) as? TimeInterval {
            let age = Date().timeIntervalSince1970 - timestamp
            let maxAge: TimeInterval = 7 * 24 * 3600 // 7 jours
            
            if age > maxAge {
                print("⏰ UserCacheManager: Cache expiré (\(Int(age/3600))h), nettoyage")
                clearCache()
                return nil
            }
        }
        
        do {
            let decoder = JSONDecoder()
            let user = try decoder.decode(AppUser.self, from: data)
            print("🚀 UserCacheManager: Utilisateur trouvé en cache: \(user.name)")
            return user
        } catch {
            print("❌ UserCacheManager: Erreur décodage cache: \(error)")
            // Nettoyer le cache corrompu
            clearCache()
            return nil
        }
    }
    
    /// Vérifie si un utilisateur est en cache
    func hasCachedUser() -> Bool {
        return userDefaults.data(forKey: cacheKey) != nil
    }
    
    /// Nettoie le cache (déconnexion, suppression compte, etc.)
    func clearCache() {
        print("🗑️ UserCacheManager: Nettoyage cache utilisateur")
        userDefaults.removeObject(forKey: cacheKey)
        userDefaults.removeObject(forKey: cacheTimestampKey)
        // Nettoyer aussi les images en cache
        clearCachedProfileImage()
        clearCachedPartnerImage()
    }
    
    // MARK: - Cache Profile Image
    
    /// Cache l'image de profil physiquement pour affichage instantané
    func cacheProfileImage(_ image: UIImage) {
        print("🖼️ UserCacheManager: Mise en cache image de profil")
        
        guard let imageData = image.jpegData(compressionQuality: 0.8) else {
            print("❌ UserCacheManager: Impossible de convertir l'image en données")
            return
        }
        
        userDefaults.set(imageData, forKey: profileImageCacheKey)
        print("✅ UserCacheManager: Image de profil mise en cache (\(imageData.count) bytes)")
    }
    
    // MARK: - Simplification: Plus besoin du système "pending"
    // Le cache local est maintenant la source de vérité pour l'affichage
    // Firebase se synchronise en arrière-plan sans affecter l'UI
    
    /// Récupère l'image de profil en cache (instantané)
    func getCachedProfileImage() -> UIImage? {
        guard let imageData = userDefaults.data(forKey: profileImageCacheKey) else {
            print("🖼️ UserCacheManager: Aucune image de profil en cache")
            return nil
        }
        
        guard let image = UIImage(data: imageData) else {
            print("❌ UserCacheManager: Impossible de charger l'image depuis le cache")
            // Nettoyer les données corrompues
            userDefaults.removeObject(forKey: profileImageCacheKey)
            return nil
        }
        
        print("✅ UserCacheManager: Image de profil trouvée en cache")
        return image
    }
    
    /// Vérifie si une image de profil est en cache
    func hasCachedProfileImage() -> Bool {
        return userDefaults.data(forKey: profileImageCacheKey) != nil
    }
    
    /// Nettoie seulement l'image de profil en cache
    func clearCachedProfileImage() {
        print("🗑️ UserCacheManager: Nettoyage image de profil en cache")
        userDefaults.removeObject(forKey: profileImageCacheKey)
    }
    
    // MARK: - Cache Partner Image
    
    /// Cache l'image de profil du partenaire avec son URL pour comparaison future
    func cachePartnerImage(_ image: UIImage, url: String) {
        print("🤝 UserCacheManager: Mise en cache image partenaire")
        
        guard let imageData = image.jpegData(compressionQuality: 0.8) else {
            print("❌ UserCacheManager: Impossible de convertir l'image partenaire en données")
            return
        }
        
        userDefaults.set(imageData, forKey: partnerImageCacheKey)
        userDefaults.set(url, forKey: partnerImageURLKey)
        // Log sécurisé sans exposer l'URL avec token
        print("✅ UserCacheManager: Image partenaire mise en cache (\(imageData.count) bytes)")
    }
    
    /// Récupère l'image du partenaire en cache
    func getCachedPartnerImage() -> UIImage? {
        guard let imageData = userDefaults.data(forKey: partnerImageCacheKey) else {
            print("🤝 UserCacheManager: Aucune image partenaire en cache")
            return nil
        }
        
        guard let image = UIImage(data: imageData) else {
            print("❌ UserCacheManager: Impossible de charger l'image partenaire depuis le cache")
            clearCachedPartnerImage()
            return nil
        }
        
        print("✅ UserCacheManager: Image partenaire trouvée en cache")
        return image
    }
    
    /// Vérifie si l'URL du partenaire a changé (pour savoir si on doit recharger)
    func hasPartnerImageChanged(newURL: String?) -> Bool {
        let cachedURL = userDefaults.string(forKey: partnerImageURLKey)
        let hasChanged = cachedURL != newURL
        
        if hasChanged {
            // Log sécurisé sans exposer les URLs avec tokens
            print("🔄 UserCacheManager: URL partenaire changée")
        } else {
            // Log sécurisé sans exposer les URLs avec tokens
            print("✅ UserCacheManager: URL partenaire inchangée")
        }
        
        return hasChanged
    }
    
    /// Vérifie si une image partenaire est en cache
    func hasCachedPartnerImage() -> Bool {
        return userDefaults.data(forKey: partnerImageCacheKey) != nil
    }
    
    /// Nettoie l'image partenaire en cache
    func clearCachedPartnerImage() {
        print("🗑️ UserCacheManager: Nettoyage image partenaire en cache")
        userDefaults.removeObject(forKey: partnerImageCacheKey)
        userDefaults.removeObject(forKey: partnerImageURLKey)
    }
    
    /// Met à jour seulement certains champs sans remplacer tout le cache
    func updateCachedUserField<T: Codable>(_ keyPath: WritableKeyPath<AppUser, T>, value: T) {
        guard var cachedUser = getCachedUser() else {
            print("⚠️ UserCacheManager: Impossible de mettre à jour - pas d'utilisateur en cache")
            return
        }
        
        cachedUser[keyPath: keyPath] = value
        cacheUser(cachedUser)
    }
    
    // MARK: - Cache Info & Recovery
    
    func getCacheInfo() -> (hasCache: Bool, lastUpdated: Date?, userCount: Int) {
        let hasCache = hasCachedUser()
        var lastUpdated: Date?
        
        if let timestamp = userDefaults.object(forKey: cacheTimestampKey) as? TimeInterval {
            lastUpdated = Date(timeIntervalSince1970: timestamp)
        }
        
        return (hasCache: hasCache, lastUpdated: lastUpdated, userCount: hasCache ? 1 : 0)
    }
    
    // MARK: - Plus besoin de détection d'incohérence
    // Avec la nouvelle approche, le cache local est toujours la source de vérité
    // Firebase se synchronise en arrière-plan sans impact sur l'expérience utilisateur
}

File: Services/UserCacheManager.kt
package com.love2loveapp.services.cache

import android.content.Context
import android.content.SharedPreferences
import android.graphics.Bitmap
import android.graphics.BitmapFactory
import android.util.Base64
import android.util.Log
import com.google.gson.Gson
import java.io.ByteArrayOutputStream
import java.util.Date
import com.love2loveapp.model.AppConstants

/**
 * Équivalent Kotlin/Android du UserCacheManager (Swift).
 *
 * ✅ JSON de l'utilisateur en SharedPreferences (Gson)
 * ✅ Images de profil & partenaire en Base64 (JPEG, qualité 0.8) dans SharedPreferences
 * ✅ TTL du cache utilisateur : 7 jours
 * ✅ Helpers: hasCache, clearCache, updateCachedUser (transform)
 *
 * ⚠️ Remarque performance: pour des images lourdes, préférer un cache disque (ex: fichiers dans cacheDir)
 *    ou une lib (Coil/Glide). Cette implémentation reste 1:1 avec l’approche UserDefaults de la version iOS.
 *
 * Utilisation :
 *   UserCacheManager.init(appContext)
 *   UserCacheManager.cacheUser(user)
 *   val cached = UserCacheManager.getCachedUser()
 */
object UserCacheManager {

    private const val TAG = "UserCacheManager"
    private const val PREFS_NAME = "cached_user_prefs"

    // Keys
    private const val KEY_USER_JSON = "cached_user_data"
    private const val KEY_USER_TS = "cached_user_timestamp"
    private const val KEY_PROFILE_IMG = "cached_profile_image"
    private const val KEY_PARTNER_IMG = "cached_partner_image"
    private const val KEY_PARTNER_IMG_URL = "cached_partner_image_url"

    // TTL 7 jours
    private const val CACHE_MAX_AGE_MS = AppConstants.Cache.USER_CACHE_MAX_AGE_MS

    private lateinit var prefs: SharedPreferences
    private val gson by lazy { Gson() }

    @Volatile private var initialized = false

    /** À appeler une seule fois (ex: dans Application.onCreate()) */
    fun init(context: Context) {
        if (!initialized) {
            synchronized(this) {
                if (!initialized) {
                    prefs = context.applicationContext.getSharedPreferences(PREFS_NAME, Context.MODE_PRIVATE)
                    initialized = true
                    Log.d(TAG, "✅ init: SharedPreferences initialisés")
                }
            }
        }
    }

    // --- Sécurité interne ---
    private fun requireInit() {
        check(initialized) { "UserCacheManager.init(context) doit être appelé avant usage." }
    }

    // ============
    // USER CACHE
    // ============

    /** Cache l'utilisateur après chargement Firebase */
    fun cacheUser(user: AppUser) {
        requireInit()
        return try {
            Log.d(TAG, "💾 Mise en cache utilisateur: ${user.name}")
            val json = gson.toJson(user)
            prefs.edit()
                .putString(KEY_USER_JSON, json)
                .putLong(KEY_USER_TS, System.currentTimeMillis())
                .apply()
            Log.d(TAG, "✅ Utilisateur mis en cache avec succès")
        } catch (t: Throwable) {
            Log.e(TAG, "❌ Erreur encodage utilisateur: ${t.message}", t)
        }
    }

    /** Récupère l'utilisateur en cache (instantané, TTL 7j) */
    fun getCachedUser(): AppUser? {
        requireInit()
        val json = prefs.getString(KEY_USER_JSON, null) ?: run {
            Log.d(TAG, "📂 Aucun utilisateur en cache")
            return null
        }

        // Vérifier l'âge du cache
        prefs.getLong(KEY_USER_TS, 0L).let { ts ->
            if (ts > 0) {
                val age = System.currentTimeMillis() - ts
                if (age > CACHE_MAX_AGE_MS) {
                    Log.d(TAG, "⏰ Cache expiré (${age / 3600000}h), nettoyage")
                    clearCache()
                    return null
                }
            }
        }

        return try {
            val user = gson.fromJson(json, AppUser::class.java)
            Log.d(TAG, "🚀 Utilisateur trouvé en cache: ${user.name}")
            user
        } catch (t: Throwable) {
            Log.e(TAG, "❌ Erreur décodage cache: ${t.message}. Nettoyage…", t)
            clearCache()
            null
        }
    }

    /** Vérifie si un utilisateur est en cache */
    fun hasCachedUser(): Boolean {
        requireInit()
        return prefs.contains(KEY_USER_JSON)
    }

    /** Nettoie le cache (déconnexion, suppression compte, etc.) */
    fun clearCache() {
        requireInit()
        Log.d(TAG, "🗑️ Nettoyage cache utilisateur")
        prefs.edit()
            .remove(KEY_USER_JSON)
            .remove(KEY_USER_TS)
            .apply()
        clearCachedProfileImage()
        clearCachedPartnerImage()
    }

    // =================
    // PROFILE IMAGE
    // =================

    /** Cache l'image de profil (JPEG 80%) */
    fun cacheProfileImage(bitmap: Bitmap) {
        requireInit()
        try {
            Log.d(TAG, "🖼️ Mise en cache image de profil")
            val bytes = bitmap.toJpegBytes(quality = AppConstants.Cache.IMAGE_CACHE_QUALITY)
            prefs.edit().putString(KEY_PROFILE_IMG, bytes.base64()).apply()
            Log.d(TAG, "✅ Image de profil mise en cache (${bytes.size} bytes)")
        } catch (t: Throwable) {
            Log.e(TAG, "❌ Impossible de convertir l'image de profil: ${t.message}", t)
        }
    }

    /** Récupère l'image de profil en cache */
    fun getCachedProfileImage(): Bitmap? {
        requireInit()
        val b64 = prefs.getString(KEY_PROFILE_IMG, null) ?: run {
            Log.d(TAG, "🖼️ Aucune image de profil en cache")
            return null
        }
        return try {
            val bytes = Base64.decode(b64, Base64.DEFAULT)
            BitmapFactory.decodeByteArray(bytes, 0, bytes.size) ?: run {
                Log.e(TAG, "❌ Decode image profil échoué, nettoyage clé")
                prefs.edit().remove(KEY_PROFILE_IMG).apply()
                null
            }.also {
                if (it != null) Log.d(TAG, "✅ Image de profil trouvée en cache")
            }
        } catch (t: Throwable) {
            Log.e(TAG, "❌ Erreur chargement image profil: ${t.message}", t)
            prefs.edit().remove(KEY_PROFILE_IMG).apply()
            null
        }
    }

    /** Présence image de profil */
    fun hasCachedProfileImage(): Boolean {
        requireInit()
        return prefs.contains(KEY_PROFILE_IMG)
    }

    /** Nettoie seulement l'image de profil en cache */
    fun clearCachedProfileImage() {
        requireInit()
        Log.d(TAG, "🗑️ Nettoyage image de profil en cache")
        prefs.edit().remove(KEY_PROFILE_IMG).apply()
    }

    // =================
    // PARTNER IMAGE
    // =================

    /** Cache l'image du partenaire + son URL source (pour détection de changement) */
    fun cachePartnerImage(bitmap: Bitmap, url: String) {
        requireInit()
        try {
            Log.d(TAG, "🤝 Mise en cache image partenaire")
            val bytes = bitmap.toJpegBytes(quality = AppConstants.Cache.IMAGE_CACHE_QUALITY)
            prefs.edit()
                .putString(KEY_PARTNER_IMG, bytes.base64())
                .putString(KEY_PARTNER_IMG_URL, url)
                .apply()
            Log.d(TAG, "✅ Image partenaire mise en cache (${bytes.size} bytes)")
        } catch (t: Throwable) {
            Log.e(TAG, "❌ Impossible de convertir l'image partenaire: ${t.message}", t)
        }
    }

    /** Récupère l'image partenaire en cache */
    fun getCachedPartnerImage(): Bitmap? {
        requireInit()
        val b64 = prefs.getString(KEY_PARTNER_IMG, null) ?: run {
            Log.d(TAG, "🤝 Aucune image partenaire en cache")
            return null
        }
        return try {
            val bytes = Base64.decode(b64, Base64.DEFAULT)
            BitmapFactory.decodeByteArray(bytes, 0, bytes.size) ?: run {
                Log.e(TAG, "❌ Decode image partenaire échoué, nettoyage clés")
                clearCachedPartnerImage()
                null
            }.also {
                if (it != null) Log.d(TAG, "✅ Image partenaire trouvée en cache")
            }
        } catch (t: Throwable) {
            Log.e(TAG, "❌ Erreur chargement image partenaire: ${t.message}", t)
            clearCachedPartnerImage()
            null
        }
    }

    /** Vérifie si l’URL partenaire a changé (pour décider si on recharge) */
    fun hasPartnerImageChanged(newURL: String?): Boolean {
        requireInit()
        val cached = prefs.getString(KEY_PARTNER_IMG_URL, null)
        val changed = cached != newURL
        if (changed) Log.d(TAG, "🔄 URL partenaire changée")
        else Log.d(TAG, "✅ URL partenaire inchangée")
        return changed
    }

    /** Présence image partenaire */
    fun hasCachedPartnerImage(): Boolean {
        requireInit()
        return prefs.contains(KEY_PARTNER_IMG)
    }

    /** Nettoie l’image partenaire */
    fun clearCachedPartnerImage() {
        requireInit()
        Log.d(TAG, "🗑️ Nettoyage image partenaire en cache")
        prefs.edit()
            .remove(KEY_PARTNER_IMG)
            .remove(KEY_PARTNER_IMG_URL)
            .apply()
    }

    // ==========================
    // UPDATE PARTIEL UTILISATEUR
    // ==========================

    /**
     * Met à jour l’utilisateur en cache via une transformation immuable.
     * Exemple :
     *   UserCacheManager.updateCachedUser { it.copy(name = "Nouveau nom") }
     */
    fun updateCachedUser(transform: (AppUser) -> AppUser) {
        requireInit()
        val current = getCachedUser()
        if (current == null) {
            Log.w(TAG, "⚠️ Impossible de mettre à jour - pas d'utilisateur en cache")
            return
        }
        val updated = try {
            transform(current)
        } catch (t: Throwable) {
            Log.e(TAG, "❌ Erreur transform utilisateur: ${t.message}", t)
            return
        }
        cacheUser(updated)
    }

    // ===========
    // INFO CACHE
    // ===========

    data class CacheInfo(
        val hasCache: Boolean,
        val lastUpdated: Date?,
        val userCount: Int
    )

    fun getCacheInfo(): CacheInfo {
        requireInit()
        val hasCache = hasCachedUser()
        val ts = prefs.getLong(KEY_USER_TS, 0L)
        val lastUpdated = if (ts > 0) Date(ts) else null
        return CacheInfo(hasCache, lastUpdated, if (hasCache) 1 else 0)
    }

    // ====================
    // Helpers internes
    // ====================

    private fun Bitmap.toJpegBytes(quality: Int = 80): ByteArray {
        val bos = ByteArrayOutputStream()
        this.compress(Bitmap.CompressFormat.JPEG, quality.coerceIn(0, 100), bos)
        return bos.toByteArray()
    }

    private fun ByteArray.base64(): String =
        Base64.encodeToString(this, Base64.DEFAULT)
}

/**
 * Ton modèle Kotlin doit être sérialisable par Gson.
 * Exemple minimal :
 *
 * data class AppUser(
 *   val id: String,
 *   val name: String,
 *   val email: String? = null,
 *   // ... autres champs (doivent rester compatibles JSON)
 * )
 */
data class AppUser(
    val id: String,
    val name: String,
    val email: String? = null
    // ajoute ici les champs dont tu as besoin
)
